### ISSUE TITLE
Invoking remote commands into WSL (bash) fails

### ISSUE THREAD

**Issue by darkverbito** [Microsoft/WSL#2322]
Open SSH 0.0.17.0
Windows 10 Enterprise N
Any SSH client (Mac OS SSH, PLINK)

I'm having success using SSH as follows:

bobby {~/code/ssh} ➩ ssh -i sshkey bobby@40.116.167.230
Microsoft Windows [Version 10.0.14393]
(c) 2016 Microsoft Corporation. All rights reserved.

bobby@WINDOWS-TST C:\Users\bobby>
bobby@WINDOWS-TST C:\Users>ls -l
total 37
lrwxrwxrwx 1 bobby 197121 14 Jul 16 2016 'All Users' -> /c/ProgramData
drwxr-xr-x 1 bobby 197121 0 Aug 22 2016 Default
lrwxrwxrwx 1 bobby 197121 16 Jul 16 2016 'Default User' -> /c/Users/Default
drwxr-xr-x 1 bobby 197121 0 Aug 22 2016 Public
drwxr-xr-x 1 bobby 197121 0 Jun 12 20:49 SSHD
drwxr-xr-x 1 bobby 197121 0 Aug 22 2016 defaultuser0
-rw-r--r-- 1 bobby 197121 174 Jul 16 2016 desktop.ini
drwxr-xr-x 1 bobby 197121 0 Jul 12 18:14 bobby

And I can execute remote commands:

bobby {~/code/ssh} ➩ ssh -i sshkey bobby@40.116.167.230 "cd .. & ls -l"
total 37
lrwxrwxrwx 1 bobby 197121 14 Jul 16 2016 All Users -> /c/ProgramData
drwxr-xr-x 1 bobby 197121 0 Aug 22 2016 Default
lrwxrwxrwx 1 bobby 197121 16 Jul 16 2016 Default User -> /c/Users/Default
drwxr-xr-x 1 bobby 197121 0 Aug 22 2016 Public
drwxr-xr-x 1 bobby 197121 0 Jun 12 20:49 SSHD
drwxr-xr-x 1 bobby 197121 0 Aug 22 2016 defaultuser0
-rw-r--r-- 1 bobby 197121 174 Jul 16 2016 desktop.ini
drwxr-xr-x 1 bobby 197121 0 Jul 12 18:14 bobby

However when I try to execute things in bash I get an error:

bobby {~/code/ssh} ➩ ssh -i sshkey bobby@40.116.167.230 "bash -c 'eval echo hi mom'"
Error: 0x80070057

But that does work if I execute it interactively:

bobby {~/code/ssh} ➩ ssh -i sshkey bobby@40.116.167.230
Microsoft Windows [Version 10.0.14393]
(c) 2016 Microsoft Corporation. All rights reserved.

bobby@WINDOWS-TST C:\Users\bobby>bash -c 'eval echo hi mom'
hi mom

I had to set Local Launch and Activation permissions for my user by taking control of the registry keys:

HKEY_LOCAL_MACHINE\SOFTWARE\Classes\AppID{F72671A9-012C-4725-9D2F-2A4D32D65169}
HKEY_CLASSES_ROOT\CLSID{4f476546-b412-4579-b64c-123df331e3d6}

Then adjusting the permissions in DCOM component services for AppId F72671A9-012C-4725-9D2F-2A4D32D65169. Before doing that I was getting an error 0x80070005 any time I would try to invoke bash over OpenSSH-Win64

**Comment by sunjoong** [comment]
@darkverbito - It looks like #2.

**Comment by benhillis** [comment]
@sunjoong - Nope this is different.  As @darkverbito points out our COM interface is locked town to interactive user.  We have an internal thread about this, let me bump that thread to see if this is something we plan to fix.

**Comment by darkverbito** [comment]
@benhillis is there any adjustment to COM that I can make to resolve the 0x80070057 error? It seems like #2 is tracking the issue where bash is invoked with -c.

**Comment by benhillis** [comment]
@darkverbito - I don't want to provide a workaround until we've decided internally if this is something we want to support.

**Comment by sunjoong** [comment]
@benhillis - I see, but you had marked #386 as duplicate of #2 so I thought this was similar.

**Comment by benhillis** [comment]
Looks like that was incorrect, I'll close that as a duplicate of this issue since this has a bit more detail.

**Comment by sunjoong** [comment]
@benhillis - Good. By the way, #2 is still marked as "fixinboud" and the last comment said "Creators update is coming soon".

**Comment by sunilmut** [comment]
Cross-linking. This is the same as PowerShell/Win32-OpenSSH#808

@darkverbito - It's a reasonable ask. We need to run this with our security counterparts and we will get back to you. Thanks for bringing this up.

**Comment by gravcat** [comment]
Can we re-open the issue for proper tracking? I'd like to know what the outcome is, and if it's waiting on internal discussion it might not be appropriate to have it in a closed state. Thoughts?

**Comment by sunilmut** [comment]
@gravcat - Thanks for your feedback. I think the reason it was closed in Win32-OpenSSH repo is because the issue is specific to WSL (see [comment](https://github.com/PowerShell/Win32-OpenSSH/issues/808#issuecomment-314879577)). I agree that the issue should remain open in both places as it manifests itself within the Win32-OpenSSH (even though it is bash specific). Please feel free to suggest that here PowerShell/Win32-OpenSSH#808

**Comment by arcnmx** [comment]
@darkverbito could you specify the settings that need to be changed in order to work around error 0x80070005 ? I wasn't able to get interactive mode to work at all, but I may be doing something wrong with the DCOM component services settings.

**Comment by benhillis** [comment]
Support for running WSL from non-interactive sessions is coming to an Insider build soon.

**Comment by gravcat** [comment]
That will be stellar @benhillis. Will this also solve the inability to launch `lxrun` from the SYSTEM context? Ideally it'd be great to invoke this from Azure Custom Script Extension and have it be able to scripted and completely autonomous.

@arcnmx I'll link you a blog post soon on it to walk through the required security settings to be able to do it _today_. It's definitely a hack but it works.

**Comment by gravcat** [comment]
@arcnmx and @benhillis (if you're interested in the run-down)

Here's a compiled list based on my notes which should reflect what's necessary today to get bash working from a less than conventional session like Win32-OpenSSH.

**Comment by benhillis** [comment]
@gravcat - Thanks for that.  It should be noted that even with your workaround some things just will not work without my fixes.  For example if a user is logged in locally and running bash, then signs in remotely via ssh and launches bash it will fail.  With my pending change that will work as expected.

**Comment by gravcat** [comment]
That's awesome! I know these things can be thankless more often than not, so I want to express my extreme gratitude for the efforts you're putting forth. 

We have been hacking on WSL for the better part of six months and these improvements will be significant for us! Hopefully so much so that we can automate the entire process and do away with having to maintain an image very soon. Cheers. :beer:

**Comment by arcnmx** [comment]
Thanks, I think I was starting from the wrong CLSID. [Your scripted version](https://gist.github.com/gravcat/4ceb977916c5f551cae44375296ab72a) is particularly helpful!

**Comment by benhillis** [comment]
@gravcat - Thank you for the kind words.  I am very grateful for our GitHub community, they have been instrumental to our success.

**Comment by benhillis** [comment]
This is fixed in 17046 and later.
